wordpress-mobile/WordPress-iOS
# Post Reblogging - Users with no available sites and users with more than one available site.

### This PR extends the post reblogging feature, covering two additional use cases
- #### users with no available site
- #### users with more than one available site

Ref: #12960

### Description
This PR extends the reblog feature to all users, covering the two use cases listed above.
The site picker will be presented to users with more than one site when they tap on `reblog`, so they can choose the site they want to reblog the selected post.
Users with no available site will be presented a `NoResultsViewController` that will give them the opportunity to be redirected to `My Sites` where they can add a new site.
The flow reflects the design below:

<img width="856" alt="version2" src="https://user-images.githubusercontent.com/34376330/70959626-62753400-2042-11ea-9193-620c6eab8ab8.png">

as discussed in the design discovery for this use case, the following subtitle has been added to NoResultsViewController

> Once you create a site, you can reblog content that you like to your own site

This PR also adds the selection of the editor (classic editor or block editor) based on the preference set by the user: the editor will be either classic or block, depending on what the user has set on the current site. It will always be possible to switch between the two, as long as the block editor is available.

The videos below illustrate how the feature works

#### (left) reblogging a post - (right) choose site

![Publish](https://user-images.githubusercontent.com/34376330/70959932-22fb1780-2043-11ea-86f4-ed730aa5ddba.gif)      ![MultipleSites](https://user-images.githubusercontent.com/34376330/70959967-41f9a980-2043-11ea-8af9-c3fa015eccb9.gif)

### 


#### (left) choose editor - (right) no available site

![SelectEditorType](https://user-images.githubusercontent.com/34376330/70959989-5178f280-2043-11ea-89f8-190232d0d2d2.gif) ![NoSites](https://user-images.githubusercontent.com/34376330/70960006-5d64b480-2043-11ea-99ca-c6678045a67f.gif)
#### Note: analytics are not included in this PR.

### Change log
- `ReblogPresenter` now contains:
  - a `presentEditor(_ :, _ :, _ : )` method, that will present `EditPostViewController` injected with the reblog post content. Depending if the user has one or more sites, the blog picker will be displayed or not
  - a `presentNoSiteScene( _ : )` method, that will present the no sites flow
- `ReblogFormatter` has been partially refactored to allow formatting content for either Classic or Block editor
- Since `ReaderReblogAction.swift` was increasing in size, has been broken up into three files:
  - `ReaderReblogAction.swift`: contains the main Type (and will have an extension for analytics when it will be added)
  - `ReaderReblogAction+Formatter`: contains the formatter
  - `ReaderReblogAction+Presenter`: contains the presenter and the `Post` extension
- A boolean parameter `postIsReblogged` has been added to `PostEditor`. As a consequence, both `AztecPostViewController` and `GutenbergViewController` now implement this parameter, defaulting it to `false`. This parameter is used to determine if the post comes from a reblogging or not. If it does, the content will be kept unaltered when switching between sites, as opposed to a normal draft post, that (on Aztec) loses its media content, and strips media-related html (like `<img src= ...`) when the destination site changes.
- Added two boolean parameters to `EditPostViewController`:
  - `openWithBlogSelector`: tells the editor if it must present the blog selector
  - `postIsReblogged`: is used to tell `PostEditor` (`Aztec` or `Gutenberg`) if the post is reblogged
- Added two optional handlers to `NoResultsViewController:
   - actionButtonHandler: (() -> Void)?
  -  dismissButtonHandler: (() -> Void)?
  - These are injected in the related action buttons. There is already a `NoResultsViewControllerDelegate` for this purpose, but since `ReaderReblogAction()` acts as `fire-and-forget`, conforming `ReaderReblogAction()` to `NoResultsViewControllerDelegate` would not work, as the reference would not hold (ReaderReblogAction() has already been deallocated when the `action` or `dismiss` buttons are tapped). Other options could be:
      - Hold a reference to `ReaderReblogAction()` in the UIViewController of origin (either `ReaderStreamViewController` (by means of `ReaderPostCellAction`) or `ReaderDetailViewController`
      - Conform the aforementioned  `UIViewControllers` to `NoResultsViewControllerDelegate`: these two already conform to that protocol and have logic for it, so it would be necessary to apply conditional logic
  - Since these two options seem more complicate, I choose the handlers, as they do not interfere with the existing logic and it's easy to understand what they do.

### To test:
1. Make sure you have two test users:
    - one with more available sites
    - one with no available sites
2. Build & Run
3. Log in with the one that has multiple sites (update `wpcom_app_credentials` as needed)
4. Go to `My Sites` and in the `Settings` turn on `Use Block Editor`
5. Go to the `Reader`, select a site of your preference from the `STREAMS` section and tap the `reblog` button on a post (alternatively, enter the post detail and tap the `reblog` button)
6. Check that the blog selector is displayed
7. Select a site (or hit cancel) and check that the block editor is shown
8. Recall the blog selector tapping on the blog name (upper left), and switch between available sites.
9. Check that the post content remains unaltered
10. Find the `switch to classic editor` in the `more` (ellipsis) menu in the upper right corner
11. Verify that the classic editor is displayed with the same content
12. Publish the post and verify that the `view` toast is displayed
13. Go back to `My Sites`, turn off `Use Block Editor`
14. Repeat step 5 and check that the classic editor is displayed now
15. Repeat steps 7-12 (or a subset of it). For the current site, the `switch to block editor` option should not be available
16. Switch to the user with no available sites (update `wpcom_app_credentials` as needed)
17. Go to the `Reader` and, either from the post list or the post detail, tap on `reblog`
18. Check that a `NoResultsViewController` is displayed, with
  - title: `No available sites`
  - subtitle: `Once you create a site, you can reblog content that you like to your own site`
  - action button label: `Manage sites`
19. Tap on `Manage sites` and check that you get redirected to `My Sites`
20. Check that `My Sites` displays the typical `NoResultsViewController` for users with no sites (see design)
21. Tap the `Add new site` button and check that the options for creating a new `WordPress.com` site or adding an existing one are available in the action sheet.

#### PR submission checklist:

  - Unit tests have been added.
  - Since it's a WIP feature, it should not be necessary.
